CPP: Implement preprocessor comments on #if directives, #error and #warning

The following MRE demonstrates several annoying preprocessor misbehaviors in LFortran 0.63:

```C
# /* this is a CPP comment */
#if __LFORTRAN__ /* LFortran */ || defined(__flang__) /* LLVM Flang */ || __GFORTRAN__ /* GNU Gfortran */
#warning Successful warning! This warning message should be visible in a compiler diagnostic.
# /* the next line should trigger a compile error with user-provided message: */
#error Success! If you are seeing this error message, the test passed.
#else  /* another comment */
#endif /* and another */
```

Demonstration:
```
cgpu$ flang cpp-mre.F90
error: Could not scan cpp-mre.F90
./cpp-mre.F90:3:2: warning: #warning Successful warning! This warning message should be visible in a compiler diagnostic.
  #warning Successful warning! This warning message should be visible in a compiler diagnostic.
   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
./cpp-mre.F90:5:2: error: #error Success! If you are seeing this error message, the test passed.
  #error Success! If you are seeing this error message, the test passed.
   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

cgpu$ gfortran cpp-mre.F90      
cpp-mre.F90:3:2:

    3 | #warning Successful warning! This warning message should be visible in a compiler diagnostic.
      |  1~~~~~~
Warning: #warning Successful warning! This warning message should be visible in a compiler diagnostic. [-Wcpp]
cpp-mre.F90:5:2:

    5 | #error Success! If you are seeing this error message, the test passed.
      |  1~~~~
Error: #error Success! If you are seeing this error message, the test passed.

cgpu$ lfortran --cpp cpp-mre.F90 
C preprocessor error: Unexpected token '*' in factor()
 --> cpp-mre.F90:2:19
  |
2 | #if __LFORTRAN__ /* LFortran */ || defined(__flang__) /* LLVM Flang */ || __GFORTRAN__ /* GNU Gfortran */
  |                   ^^ 


Note: Please report unclear, confusing or incorrect messages as bugs at
https://github.com/lfortran/lfortran/issues.

cgpu$ lfortran --version
LFortran version: 0.63.0
Status: alpha (expected to fail on third-party codes)
Platform: Linux
LLVM: 22.1.0
Default target: x86_64-unknown-linux-gnu
```

After deleting the problematic comments on line 2, the lfortran compilation then produces:
```
cgpu$ lfortran --cpp cpp-mre2.F90
warning: Unsupported macro
 --> cpp-mre2.F90:3:1
  |
3 | #warning Successful warning! This warning message should be visible in a compiler diagnostic.
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Ignored

warning: Unsupported macro
 --> cpp-mre2.F90:4:2
  |
4 | # /* the next line should trigger a compile error with user-provided message: */
  |  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Ignored


Note: Please report unclear, confusing or incorrect messages as bugs at
https://github.com/lfortran/lfortran/issues.
/usr/bin/ld: /lib/x86_64-linux-gnu/Scrt1.o: in function `_start':
(.text+0x1b): undefined reference to `main'
clang: error: linker command failed with exit code 1 (use -v to see invocation)
...
```
The second seems to be an unimplemented case left-over from #11069.

After deleting those two lines, the lfortran compilation then produces:
```
cgpu$ lfortran --cpp cpp-mre3.F90  
warning: Unsupported macro
 --> cpp-mre3.F90:3:1
  |
3 | #error Success! If you are seeing this error message, the test passed.
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Ignored


Note: Please report unclear, confusing or incorrect messages as bugs at
https://github.com/lfortran/lfortran/issues.
/usr/bin/ld: /lib/x86_64-linux-gnu/Scrt1.o: in function `_start':
(.text+0x1b): undefined reference to `main'
clang: error: linker command failed with exit code 1 (use -v to see invocation)
...
```

The `#error` directive dates back to C89, and by widespread convention should cause translation to fail (not proceed to linking). 

`#warning` is new in C23, but widely implemented and also worth supporting.

--- a/integration_tests/CMakeLists.txt
+++ b/integration_tests/CMakeLists.txt
@@ -3792,6 +3792,10 @@ RUN(NAME cpp_pre_15 LABELS gfortran llvm
     EXTRA_ARGS --cpp
     GFORTRAN_ARGS -cpp)
 
+RUN(NAME cpp_pre_16 LABELS gfortran llvm
+    EXTRA_ARGS --cpp
+    GFORTRAN_ARGS -cpp)
+
 RUN(NAME dabs_01 LABELS gfortran llvmImplicit)
 
 RUN(NAME minpack_03 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc)

--- a/integration_tests/cpp_pre_16.f90
+++ b/integration_tests/cpp_pre_16.f90
@@ -1,0 +1,20 @@
+program cpp_pre_16
+    implicit none
+    integer :: x
+
+#if 1 /* taken */ || defined(__NOTDEFINED_MACRO__) /* not defined */
+    x = 42
+#else
+#error This branch must not be selected
+#endif
+
+#if 0 /* not taken */
+#error This #error must be ignored because branch is not taken
+#warning This #warning must be ignored because branch is not taken
+#endif
+
+# /* a CPP null directive with a C-style comment */
+
+    if (x /= 42) error stop
+    print *, x
+end program

--- a/src/lfortran/parser/preprocessor.re
+++ b/src/lfortran/parser/preprocessor.re
@@ -612,6 +612,30 @@ Result<std::string> CPreprocessor::run(const std::string &input, LocationManager
                 interval_end_type_0(lm, output.size(), cur-string_start);
                 continue;
             }
+            "#" whitespace? "warning" [ \t\v\r]* @t1 [^\n\x00]* @t2 newline {
+                if (!branch_enabled) continue;
+                std::string msg = token(t1, t2);
+                Location loc;
+                loc.first = tok - string_start;
+                loc.last = (cur > string_start ? cur - 1 : cur) - string_start;
+                diagnostics.add(diag::Diagnostic(
+                    "#warning " + msg, diag::Level::Warning,
+                    diag::Stage::CPreprocessor,
+                    { diag::Label("", {loc}) }));
+                interval_end_type_0(lm, output.size(), cur-string_start);
+                continue;
+            }
+            "#" whitespace? "error" [ \t\v\r]* @t1 [^\n\x00]* @t2 newline {
+                if (!branch_enabled) continue;
+                std::string msg = token(t1, t2);
+                Location loc;
+                loc.first = tok - string_start;
+                loc.last = (cur > string_start ? cur - 1 : cur) - string_start;
+                throw PreprocessorError(diag::Diagnostic(
+                    "#error " + msg, diag::Level::Error,
+                    diag::Stage::CPreprocessor,
+                    { diag::Label("", {loc}) }));
+            }
             "#" whitespace? "include" whitespace ["<] @t1 [^">\x00]* @t2 [">] [^\n\x00]* newline {
                 if (!branch_enabled) continue;
                 std::string filename = token(t1, t2);
@@ -996,6 +1020,8 @@ void get_next_token(unsigned char *string_start, unsigned char *&cur, CPPTokenTy
             end { type = CPPTokenType::TK_EOF; return; }
             newline { type = CPPTokenType::TK_EOF; return; }
             whitespace { continue; }
+            "//" [^\n\x00]* { continue; }
+            "/*" ([^*\x00] | "*"[^/\x00])* "*/" { continue; }
             "\\" whitespace? newline { continue; }
             "+" { type = CPPTokenType::TK_PLUS; return; }
             "-" { type = CPPTokenType::TK_MINUS; return; }

